### PR TITLE
NetSim: Router stats

### PR DESCRIPTION
**Pull request by islemaster**
Live router statistics display on the router tab.  We're not necessarily done adding stats, but I want to get this into staging as a starting point for working with the ed team tomorrow.

out

Provides the following stats:
- Uptime: Just time since router creation
- Queued Packets: How many messages are waiting to be sent by the router
- Memory use: Total size of the messages waiting to be sent by the router (out of total available memory)
- Throughput: A little tricky... how many bits we anticipate the router will process in the next second, out of its maximum bandwidth.  I tried doing this looking backwards, averaging over router logs, that kind of thing, but this seems to be most accurate.
- Routed Packets: All-time total successfully routed packets out of total routing attempts (so, complement of drop rate)
- Routed Data: All-time total successfully routed data out of total attempted routed data.

### CODE REVIEW

File: apps/src/netsim/dataConverters.js [hunk 1]
```diff
@@ -129,21 +130,11 @@ exports.binaryToInt = function (binaryString) {
...
 
-exports.zeroPadLeft = function (string, desiredWidth) {
-  var padding = '0'.repeat(desiredWidth);
-  return (padding + string).slice(-desiredWidth);
-};
-
-exports.zeroPadRight = function (string, desiredWidth) {
-  var padding = '0'.repeat(desiredWidth);
-  return (string + padding).substr(0, desiredWidth);
-};
```
Comment (islemaster) [inline apps/src/netsim/dataConverters.js]: These moved to netsimUtils.js

File: apps/src/netsim/NetSimRouterStatsTable.js [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  return (!this.lastRenderTime_ ||
+      clock.time - this.lastRenderTime_ > MAX_RENDER_DELAY_MS);
+};
+
+/**
+ * Mark the router log data dirty, so that it will re-render on the
+ * next frame.
+ */
+NetSimRouterStatsTable.prototype.setNeedsRender = function () {
+  this.lastRenderTime_ = null;
```
Comment (breville) [inline apps/src/netsim/NetSimRouterStatsTable.js]: john carmack had a post about the dangers of overloading values with null to indicate an additional state, but i can't find it.  plus, everybody does it :)

**Comment by breville** [comment]
lgtm.  remind me to tell you sometime of the guy at xbox who wrote a complete router simulator while simultaneously writing a netstack from scratch :)

File: apps/src/netsim/NetSimRouterStatsTable.js (hunk 2 above)
Comment (islemaster) [inline apps/src/netsim/NetSimRouterStatsTable.js]: along the way I actually had a version of this with both a `needsRender` flag and a `lastRenderTime` field, but realized it was easy enough to encapsulate.

I think you're referring to [this article](http://www.gamasutra.com/view/news/128836/InDepth_Static_Code_Analysis.php) (about halfway down the page).

**Pull request merged**